src/main/java/com/arpnetworking/clusteraggregator/Main.java

```diff
@@ -245,6 +252,37 @@ private void shutdownAkka() {
     private static final Logger LOGGER = com.arpnetworking.steno.LoggerFactory.getLogger(Main.class);
     private static final Duration SHUTDOWN_TIMEOUT = Duration.create(3, TimeUnit.MINUTES);
     private static final SourceTypeLiteral SOURCE_TYPE_LITERAL = new SourceTypeLiteral();
+    private static final Semaphore SHUTDOWN_SEMAPHORE = new Semaphore(0);
+    private static final Thread SHUTDOWN_THREAD = new ShutdownThread();
+
+    private static final class ShutdownThread extends Thread {
+        private ShutdownThread() {
+            super("ClusterAggregatorShutdownHook");
+        }
+
+        @Override
+        public void run() {
+            LOGGER.info()
+                    .setMessage("Stopping cluster-aggregator")
+                    .log();
+            final int awaitingShutdownSignal = SHUTDOWN_SEMAPHORE.getQueueLength();
```

Comment:
The counting looks wonky. So there is a race condition to have all threads waiting for shutdown acquire the semaphore before the shutdown thread runs. Worse, it then assumes that they all did because it re-uses the semaphore to signal that shutdown has completed. First, if you want the bi-directional communication you need to use two constructs. Second, I don't think we need to support multiple threads (at least not yet) and if we do we should consider more carefully which construct to use - off the top of my head I think you are actually looking for a barrier.